tokio-rustls/src/lib.rs

```diff
@@ -126,6 +146,122 @@ impl TlsAcceptor {
     }
 }
 
+pub struct LazyConfigAcceptor<IO> {
+    acceptor: rustls::Acceptor,
+    buf: BytesMut,
+    used: usize,
+    io: Option<IO>,
+}
+
+impl<IO> LazyConfigAcceptor<IO>
+where
+    IO: AsyncRead + AsyncWrite + Unpin,
+{
+    #[inline]
+    pub fn new(acceptor: rustls::Acceptor, io: IO) -> Self {
+        Self {
+            acceptor,
+            buf: BytesMut::from(&[0; 512][..]),
+            used: 0,
+            io: Some(io),
+        }
+    }
+}
+
+impl<IO> Future for LazyConfigAcceptor<IO>
+where
+    IO: AsyncRead + AsyncWrite + Unpin,
+{
+    type Output = Result<Accepted<IO>, io::Error>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+        while this.acceptor.wants_read() {
+            let io = match this.io.as_mut() {
+                Some(io) => io,
+                None => {
+                    return Poll::Ready(Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        "acceptor cannot be polled after acceptance",
+                    )))
+                }
+            };
+
+            let mut buf = ReadBuf::new(&mut this.buf);
+            buf.advance(this.used);
+            if let Err(err) = ready!(Pin::new(io).poll_read(cx, &mut buf)) {
+                return Poll::Ready(Err(err));
+            }
+
+            let read = match this.acceptor.read_tls(&mut buf.filled()) {
+                Ok(read) => read,
+                Err(err) => return Poll::Ready(Err(err)),
+            };
+
+            let filled = buf.filled().len();
+            if read < filled {
+                this.buf.copy_within(read.., 0);
+                this.used = filled - read;
+            }
+
+            match this.acceptor.accept() {
+                Ok(Some(accepted)) => {
+                    let io = this.io.take().unwrap();
+                    return Poll::Ready(Ok(Accepted { accepted, io }));
+                }
+                Ok(None) => continue,
+                Err(err) => {
+                    return Poll::Ready(Err(io::Error::new(io::ErrorKind::InvalidInput, err)))
+                }
+            }
+        }
+
+        Poll::Pending
+    }
+}
+
+pub struct Accepted<IO> {
```

Comment:
Naming nit: Is this really `Accepted`? Accepted is for me the state where we accepted a TCP connection. `Connected` is when we finished the handshake. Maybe `Handshaking/Handshaking/HandshakeStart/HandshakeRequest`?

Reply:
Renamed it to `StartHandshake` for now.